### PR TITLE
Code Studio integration with Minecraft Code Connection

### PR DESCRIPTION
**Pull request by WilliamHarmonMS**

### CODE REVIEW

File: apps/src/craft/code-connection/cc-client.js [hunk 1]
```diff
@@ -4,10 +4,27 @@ import $ from 'jquery';
...
 export default class cc_client {
+
+    callbackRef = null;
+    keyRef = null;
+
     constructor(port) {
         this.port = port;
+        if (window.ipcRenderer !== undefined) {
+            window.ipcRenderer.on('responseFromApp', (event, arg) =>{
+                let response = JSON.parse(arg);
```
Comment (Hamms) [inline apps/src/craft/code-connection/cc-client.js]: nit; I'd love to see something more descriptive than "arg" for this argument

File: apps/src/craft/code-connection/cc-client.js [hunk 2]
```diff
@@ -41,21 +62,31 @@ export default class cc_client {
...
-                callback();
-            }
-        });
+        //before ajax
+        if (window.ipcRenderer !== undefined) {
+             window.ipcRenderer.sendToHost('sendToApp', {
+             'api': 'basic',
+             'url': `${baseUrl}${this.port}/${command}`
+             });
+             this.callbackRef = callback;
```
Comment (Hamms) [inline apps/src/craft/code-connection/cc-client.js]: is this assignment guaranteed to happen before `responseFromApp` events start getting sent, or do we need an `if (this.callbackRef !== undefined)` somewhere in the handler?

File: apps/src/craft/code-connection/cc-client.js [hunk 3]
```diff
@@ -18,19 +35,23 @@ export default class cc_client {
...
-                // cc responded
-                asyncCallback(data);
-            },
-            error: function (jqxhr, textStatus, error) {
-                // TODO: handle net::ERR_CONNECTION_REFUSED error gracefully
-                asyncCallback(false);
-            }
-        });
+        if (window.ipcRenderer !== undefined) {
+            asyncCallback(window.ipcRenderer !== null);
```
Comment (Hamms) [inline apps/src/craft/code-connection/cc-client.js]: are `undefined` and `null` both valid and expected values for `ipcRenderer`?

File: apps/src/craft/code-connection/cc-client.js (hunk 1 above)
Comment (WilliamHarmonMS) [reply apps/src/craft/code-connection/cc-client.js]: Roger that! It's basically the response from the Code Connection app in JSON format. I can easily make that more explicit.

File: apps/src/craft/code-connection/cc-client.js (hunk 3 above)
Comment (WilliamHarmonMS) [reply apps/src/craft/code-connection/cc-client.js]: Apparently not all browsers treat them the same. They should really be undefined. Looks like my last commit missed one!

File: apps/src/craft/code-connection/cc-client.js (hunk 2 above)
Comment (WilliamHarmonMS) [reply apps/src/craft/code-connection/cc-client.js]: That is my understanding. Sending things over the 'sendToApp' channel takes a message to Code Connection which parses the command url, converts it to a format that Minecraft understands, sends it over there, and then ends with sending a response back to Code Studio with the results of the command (such as that it completed successfully, or in some cases where a response is needed, like checking if a specific block was found, returning the result of the check so that conditionals can be properly evaluated). 

Long story short, 'responseFromApp' shouldn't ever be hit unless we're sending something to the App first.

**Comment by WilliamHarmonMS** [comment]
Also, for the style guide, I can make any nit changes that you like :)
I know I have the Hour of Code craft repo set up to run npm test and the linter will throw errors at me if something isn't in line. Does this have a similar system that could be set up to check for similar issues?

**Comment by Hamms** [comment]
Yeah! We have some [git hooks](https://github.com/code-dot-org/code-dot-org/tree/staging/tools/hooks) that lint on commits and which you can install with `rake install:hooks`

Given that most or all of your changes are going to be isolated to apps code, you could also just run [the apps-specific linting](https://github.com/code-dot-org/code-dot-org/blob/staging/tools/hooks/lint.rb#L54)

File: apps/src/craft/code-connection/cc-client.js [hunk 4]
```diff
@@ -18,19 +35,23 @@ export default class cc_client {
...
-                // cc responded
-                asyncCallback(data);
-            },
-            error: function (jqxhr, textStatus, error) {
-                // TODO: handle net::ERR_CONNECTION_REFUSED error gracefully
-                asyncCallback(false);
-            }
-        });
+        if (window.ipcRenderer !== undefined) {
+            asyncCallback(window.ipcRenderer !== undefined);
```
Comment (Hamms) [inline apps/src/craft/code-connection/cc-client.js]: note that this will now always be `true`

Comment (WilliamHarmonMS) [reply apps/src/craft/code-connection/cc-client.js]: Roger that. I think that's the state it needs to be in if we're using the ipcRenderer (I am by no means a seasoned expert with Electron apps, or the cc-client async function call). I did this work about a month ago, so the memory is a bit hazy, but I believe it wasn't working correctly without that setting. If in testing something comes up and there is bad behavior somewhere we can absolutely change stuff.

Just let me know 👍

**Comment by WilliamHarmonMS** [comment]
So with the linter aside, this PR should be looking good. I've adopted the same style guide directions that I use in the craft repo with all the code I added, and npm run lint command reports 0 warnings, but I'm not 100% sure that it's working correctly. I've run rake install:hooks but was having trouble getting the apps-specific linting to work. Unless you have any suggestions in regards to double checking with that, I'm going to push up some of the style changes (mainly indentations).

Thanks a ton :)

attached: lint running without throwing anything (at least anything apparent)

screen shot 2017-09-08 at 9 40 17 am
